### PR TITLE
Added a proof of concept for initialData making TData never undefined

### LINKED ISSUE #3310: better type narrowing for initialData
when providing `initialData` directly on a `query`, the type of `data` is guaranteed to be `TData`, _not_ `TData | undefined`:

```
const { data } = useQuery(['test'], () => Promise.resolve('test'), { initialData: 'something' })

// 🚨 Object is possibly 'undefined'.(2532)
data.toUpperCase()
```

[TypeScript playground](https://www.typescriptlang.org/play?ssl=6&ssc=19&pln=1&pc=1#code/JYWwDg9gTgLgBAbzgVwM4FMCKz1QJ5wC+cAZlBCHAORToCGAxjALQCOO+VAUFwxAHap4SACZ0YdInAC8KDNlx4AFAG0qMdEKoBdADRwlAShkA+OAAVyIYBgB0tVBAA2AN3RL1mmFUP6kwfmAYYDonABFxOgAuakcQdBgACwCAcyoiQx4Aeiy4QF4NwAq9uAB5ACMAK3QmOBs4SFRUYFKnAipkfhF0EgD0ESpbJQAmAFYAZkHMsQlbGAgAVTAwXABhOgwjIA)

ideally, we could narrow the type when `initialData` is provided. This _only_ works for `initialData` because `initialData` is executed in the constructor of the query, and because it is synchronous. Also, it can only work on `v4` because in `v3`, `undefined` is technically still valid for the `TData` generic, so we cannot remove it from the union. In `v4`, `undefined` is illegal for successful queries, so that would now work.

Keep in mind that `initialData` can also be a function that returns `TData | undefined`, so this should also narrow if the function returns `TData`.

It does _NOT_ work for:
- `placeholderData`, because that will revert back to `undefined` when a query errors
- `initialData` being provided globally, because TS cannot know that on an instance level. It needs to be explicitly passed to `useQuery`.
- `suspense` or any other case because a query can be `cancelled` while it is initially `fetching`, in which case it will be `reset` to its initial state. `queryClient.resetQueries` also does that. Also, without network connection in the default network mode, the query will be in `fetchStatus: 'paused'` and will also not have data.

I guess we could potentially do this with another overload that has `initialData` provided and removes `undefined` from the result union, but we would also have do this for `useQueries`.
I would like to avoid having to add _another_ generic to `useQuery`, as we already have four of them.

Because this is for v4, PRs should go to the `beta` branch please.

### PR DESCRIPTION
**Pull request by mattpocock**
closes #3310

### CODE REVIEW

File: packages/react-query/src/useQuery.ts [hunk 1]
```diff
@@ -15,8 +15,24 @@ export function useQuery<
...
+export function useQuery<
+  TQueryFnData = unknown,
+  TError = unknown,
+  TData = TQueryFnData,
+  TQueryKey extends QueryKey = QueryKey,
+>(
+  options: Omit<
+    UseQueryOptions<TQueryFnData, TError, TData, TQueryKey>,
+    'initialData'
+  > & { initialData: TData | (() => TData) },
```
Comment (TkDodo) [inline packages/react-query/src/useQuery.ts:33]: this needs to be `TQueryFnData` as `TData` is the type that runs through select if it exists:

```suggestion
  > & { initialData: TQueryFnData | (() => TQueryFnData) },
```

File: packages/react-query/src/useQuery.ts [hunk 2]
```diff
@@ -26,9 +42,23 @@ export function useQuery<
...
+  TQueryFnData = unknown,
+  TError = unknown,
+  TData = TQueryFnData,
+  TQueryKey extends QueryKey = QueryKey,
+>(
+  queryKey: TQueryKey,
+  options?: Omit<
+    UseQueryOptions<TQueryFnData, TError, TData, TQueryKey>,
+    'queryKey' | 'initialData'
+  > & { initialData: TData | (() => TData) },
```
Comment (TkDodo) [inline packages/react-query/src/useQuery.ts:59]:
```suggestion
  > & { initialData: TQueryFnData | (() => TQueryFnData) },
```

File: packages/react-query/src/useQuery.ts [hunk 3]
```diff
@@ -39,9 +69,24 @@ export function useQuery<
...
+  TError = unknown,
+  TData = TQueryFnData,
+  TQueryKey extends QueryKey = QueryKey,
+>(
+  queryKey: TQueryKey,
+  queryFn: QueryFunction<TQueryFnData, TQueryKey>,
+  options?: Omit<
+    UseQueryOptions<TQueryFnData, TError, TData, TQueryKey>,
+    'queryKey' | 'queryFn' | 'initialData'
+  > & { initialData: TData | (() => TData) },
```
Comment (TkDodo) [inline packages/react-query/src/useQuery.ts:87]:
```suggestion
  > & { initialData: TQueryFnData | (() => TQueryFnData) },
```

**Comment by TkDodo** [comment]
@all-contributors add @mattpocock for code

**Comment by Alcas1** [comment]
Is this going to be in V3 as well or only for V4?

**Comment by TkDodo** [comment]
@Alcas1 only on v4. please see the issue description for the reasoning behind this:
- https://github.com/TanStack/query/issues/3310

**Review by Console-buche — COMMENTED** [review]
nothing (mistake)

**Comment by moris-moze** [comment]
@TkDodo Hi, in what specific v4 version is this available? I've updated to [4.0.0-beta.23], but the `data` type still remains `T | undefined`. Thanks

**Comment by TkDodo** [comment]
just use latest v4 ?

**Comment by moris-moze** [comment]
@TkDodo Looking at https://www.npmjs.com/package/react-query/v/4.0.0-beta.23?activeTab=versions, 4.0.0-beta.23 seems to be the latest?

**Comment by TkDodo** [comment]
the package is `@tanstack/react-query`:

https://www.npmjs.com/package/@tanstack/react-query

**Comment by moris-moze** [comment]
@TkDodo Is it possible to make a query run only once, while having `initialData` set and `staleTime` set to Infinity?

E.g.:

```
export const useAllUsers = () =>
  useQuery<UserResponseDto[]>({
    queryKey: [QueryClientKeys.GET_ALL_USERS],
    queryFn: async () => {
      const response = await http.get<
        UserResponseDto[],
        AxiosResponse<UserResponseDto[]>
      >(`${API_BASE}/api/user/all`);
      return response.data;
    },
    staleTime: Infinity,
    initialData: [],
  });
```

The goal here is not to have possibly undefined data, hence the `initialData`. Besides that, I want to run the query only once so there's the `staleTime: Infinity`. This obviously can't work since query has initial data and stale time is infinite, so there is no triggering of the queryFn. Is there a way around this?

**Comment by TkDodo** [comment]
> @TkDodo Is it possible to make a query run only once, while having initialData set and staleTime set to Infinity?

maybe with `refetchOnMount: 'always'` ? But it will fetch whenever a new observer mounts

**Pull request merged**